### PR TITLE
feat: Implement ERC721CommonUpgradeable contract

### PR DESCRIPTION
**Pull request by huyhuynh3103**
### Description
Old PR: [PR#10](https://github.com/axieinfinity/contract-template/pull/10).

### Checklist
- [x] I have clearly commented on all the main functions following the [NatSpec Format](https://docs.soliditylang.org/en/v0.8.0/natspec-format.html)
- [x] The box that allows repo maintainers to update this PR is checked
- [x] I tested locally to make sure this feature/fix works

### CODE REVIEW

**Review by TuDo1403 — APPROVED** [review]
LGTM

File: src/upgradeable/ERC721PresetMinterPauserAutoIdCustomizedUpgradeable.sol [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    _mint(to, _tokenId);
+    _tokenIdTracker.increment();
+  }
+
+  /**
+   * @dev This empty reserved space is put in place to allow future versions to add new
+   * variables without shifting down storage in the inheritance chain.
+   * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+   */
+  uint256[48] private __gap;
```
Comment (ducthotran2010) [inline src/upgradeable/ERC721PresetMinterPauserAutoIdCustomizedUpgradeable.sol:178]:
```suggestion
  uint256[50] private __gap;
```

**Review by ducthotran2010 — CHANGES_REQUESTED** [review]

File: test/foundry/SampleERC721Upgradeable.t.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+  event NonceUpdated(uint256 indexed _tokenId, uint256 indexed _nonce);
+
+  string public constant NAME = "SampleERC721";
+  string public constant SYMBOL = "NFT";
+  string public constant BASE_URI = "http://example.com/";
+
+  address _proxyAdmin;
+
+  ERC721CommonUpgradeable internal _t;
```
Comment (nxqbao) [inline test/foundry/SampleERC721Upgradeable.t.sol:27]: At least write a comment

```suggestion
  ERC721CommonUpgradeable internal _t; // The test token
```

File: src/upgradeable/refs/ERC721NonceUpgradeable.sol [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * @title ERC721NonceUpgradeable
+ * @dev This contract provides a nonce that will be increased whenever the token is tranferred.
+ */
+abstract contract ERC721NonceUpgradeable is ERC721Upgradeable {
+  /// @dev Emitted when the token nonce is updated
+  event NonceUpdated(uint256 indexed _tokenId, uint256 indexed _nonce);
+
+  /// @dev Mapping from token id => token nonce
+  mapping(uint256 => uint256) public nonces;
```
Comment (nxqbao) [inline src/upgradeable/refs/ERC721NonceUpgradeable.sol:15]:
```suggestion
  mapping(uint256 tokenId => uint256) public nonces;
```

File: src/upgradeable/ERC721PresetMinterPauserAutoIdCustomizedUpgradeable.sol (hunk 1 above)
Comment (nxqbao) [inline src/upgradeable/ERC721PresetMinterPauserAutoIdCustomizedUpgradeable.sol:178]: bring these lines to before the first method

File: src/upgradeable/ERC721CommonUpgradeable.sol [hunk 4]
```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "../refs/IERC721State.sol";
+import "./refs/ERC721NonceUpgradeable.sol";
+import "./ERC721PresetMinterPauserAutoIdCustomizedUpgradeable.sol";
+
+abstract contract ERC721CommonUpgradeable is
```
Comment (nxqbao) [inline src/upgradeable/ERC721CommonUpgradeable.sol:8]: Convention of variables `_abcxyz`!!!!

File: src/upgradeable/refs/ERC721NonceUpgradeable.sol [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import "../../../lib/openzeppelin-contracts-upgradeable/contracts/token/ERC721/ERC721Upgradeable.sol";
+
+/**
+ * @title ERC721NonceUpgradeable
+ * @dev This contract provides a nonce that will be increased whenever the token is tranferred.
+ */
+abstract contract ERC721NonceUpgradeable is ERC721Upgradeable {
+  /// @dev Emitted when the token nonce is updated
+  event NonceUpdated(uint256 indexed _tokenId, uint256 indexed _nonce);
```
Comment (nxqbao) [inline src/upgradeable/refs/ERC721NonceUpgradeable.sol:12]: Please revise all conventions!

```suggestion
  event NonceUpdated(uint256 indexed tokenId, uint256 indexed nonce);
```

File: src/upgradeable/ERC721CommonUpgradeable.sol [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+   * @dev Override `IERC165-supportsInterface`.
+   */
+  function supportsInterface(bytes4 interfaceId)
+    public
+    view
+    virtual
+    override(ERC721Upgradeable, ERC721PresetMinterPauserAutoIdCustomizedUpgradeable)
+    returns (bool)
+  {
+    return super.supportsInterface(interfaceId);
```
Comment (nxqbao) [inline src/upgradeable/ERC721CommonUpgradeable.sol:51]: `super.method()` only invokes the right-most inherited contract’s method.
Should be something like `interfaceId == type(ERC721Upgradeable).interfaceId || interfaceId == type(ERC721PresetMinterPauserAutoIdCustomizedUpgradeable).interfaceId`

Please write a test for this! Many third parties impl this method incorrectly already.

Comment (nxqbao) [reply src/upgradeable/ERC721CommonUpgradeable.sol:51]: Please check other `super.f()` in other methods as well. Write a few tests!

File: src/upgradeable/ERC721CommonUpgradeable.sol [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  function bulkMint(address[] calldata _recipients)
+    external
+    virtual
+    onlyRole(MINTER_ROLE)
+    returns (uint256[] memory _tokenIds)
+  {
+    if (_recipients.length == 0) revert ErrInvalidArrayLength();
+    _tokenIds = new uint256[](_recipients.length);
+
+    for (uint256 _i = 0; _i < _recipients.length; _i++) {
```
Comment (nxqbao) [inline src/upgradeable/ERC721CommonUpgradeable.sol:85]: Cache `length`. `++i`.

File: src/upgradeable/ERC721PresetMinterPauserAutoIdCustomizedUpgradeable.sol [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+   */
+
+  function __ERC721PresetMinterPauserAutoId_init(string memory name, string memory symbol, string memory baseTokenURI)
+    internal
+    onlyInitializing
+  {
+    __ERC721_init_unchained(name, symbol);
+    __Pausable_init_unchained();
+    __ERC721PresetMinterPauserAutoId_init_unchained(name, symbol, baseTokenURI);
+    _tokenIdTracker.increment();
```
Comment (nxqbao) [inline src/upgradeable/ERC721PresetMinterPauserAutoIdCustomizedUpgradeable.sol:69]: Why this line is not in `_init_unchained` but here? If it `_unchained` it means no need to increase the id tracker?

File: src/upgradeable/refs/ERC721NonceUpgradeable.sol [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  /**
+   * @dev Override `ERC721Upgradeable-_beforeTokenTransfer`.
+   */
+  function _beforeTokenTransfer(address _from, address _to, uint256 _firstTokenId, uint256 _batchSize)
+    internal
+    virtual
+    override
+  {
+    for (uint256 _tokenId = _firstTokenId; _tokenId < _firstTokenId + _batchSize; _tokenId++) {
```
Comment (nxqbao) [inline src/upgradeable/refs/ERC721NonceUpgradeable.sol:31]:
```suggestion
    for (uint256 _tokenId = _firstTokenId; _tokenId < _firstTokenId + _batchSize; ++_tokenId) {
```

File: src/upgradeable/refs/ERC721NonceUpgradeable.sol [hunk 10]
```diff
@@ -1,36 +1,45 @@
...
+  function _beforeTokenTransfer(address from, address to, uint256 firstTokenId, uint256 batchSize)
     internal
     virtual
     override
   {
-    for (uint256 _tokenId = _firstTokenId; _tokenId < _firstTokenId + _batchSize; _tokenId++) {
-      emit NonceUpdated(_tokenId, ++nonces[_tokenId]);
+    uint256 length = firstTokenId + batchSize;
+    unchecked {
+      for (uint256 tokenId = firstTokenId; tokenId < length; ++tokenId) {
```
Comment (nxqbao) [inline src/upgradeable/refs/ERC721NonceUpgradeable.sol:38]: I don't have proof, just feel unsafe here.

**Review by nxqbao — APPROVED** [review]

**Pull request merged**